docs/layer2/layer2_support.md

```diff
@@ -0,0 +1,374 @@
+
+      send_user_state_event running 1000 "Configuring Network"
+
+      # Restart networking and check for failures
+      if ! systemctl restart networking; then
+        echo "Network restart failed" >&2
+        send_user_state_event failed 1002 "Network restart failed"
+        exit 1
+      fi
+
+      # Verify network configuration
+      verification_failed=false
+      {{ range $index, $vlan := .VLANs }}
+      if ip addr show {{ $vlan.PortName }}.{{ $vlan.ID }} | grep -q {{ $vlan.IPAddress }}; then
+        echo "Configuration for VLAN {{ $vlan.ID }} on {{ $vlan.PortName }} with IP {{ $vlan.IPAddress }} successful"
+      else
+        echo "Configuration for VLAN {{ $vlan.ID }} on {{ $vlan.PortName }} with IP {{ $vlan.IPAddress }} failed" >&2
+        verification_failed=true
+      fi
+      {{ end }}
+
+      if [ "$verification_failed" = true ]; then
+        send_user_state_event failed 1002 "Network configuration failed"
+        exit 1
+      else
+        send_user_state_event succeeded 1001 "Network configuration successful"
+      fi
+
+# Commands to run after the script is written
+runcmd:
+  - |
+      # Fetch metadata and set up network interfaces
+      metadata=$(curl -sf https://metadata.platformequinix.com/metadata)
+
+      # Extract MAC addresses for eth0 and eth1
+      mac_eth0=$(echo "$metadata" | jq -r '.network.interfaces[] | select(.name == "eth0") | .mac')
+      mac_eth1=$(echo "$metadata" | jq -r '.network.interfaces[] | select(.name == "eth1") | .mac')
+
+      # Function to find interface name by MAC address
+      find_interface_by_mac() {
+        local mac="$1"
+        for iface in $(ls /sys/class/net/); do
+          iface_mac=$(ethtool -P "$iface" 2>/dev/null | awk '{print $NF}')
+          if [ "$iface_mac" == "$mac" ]; then
+            echo "$iface"
+            return
+          fi
+        done
+        echo "Interface not found for MAC $mac" >&2
+        return 1
+      }
+
+      # Find interface names for eth0 and eth1
+      iface_eth0=$(find_interface_by_mac "$mac_eth0")
+      iface_eth1=$(find_interface_by_mac "$mac_eth1")
+
+      # Replace eth0 and eth1 in the script with the actual interface names
+      sed -i "s/eth0/${iface_eth0}/g" /tmp/final_configuration.sh
+      sed -i "s/eth1/${iface_eth1}/g" /tmp/final_configuration.sh
+
+      # Execute the modified script
+      bash /tmp/final_configuration.sh
+
+```
+
+The CAPP will use go-templates to substitute the placeholders with appropriate values given by the user.
+
+### Layer 2 Networking Setup by the CAPP Operator
+When provisioning a metal node with Layer 2 networking, the Cluster API Provider (CAPP) Operator will perform the following steps:
+1. **Create a ConfigMap for IP Address Management**: The operator will create a new ConfigMap named <cluster_name-port_name> for each port to manage IP addresses. This ConfigMap is critical for tracking and allocating IP addresses as detailed in the *IP Address Management* section.
```

Comment:
Wondering why do we need to create a config map per port for IP management

Reply:
@niravparikh05 @displague so the desired per-node configuration could be stored in Machine spec only.
For assignging IP addresses at per machine-port level, I was thnking to use ConfigMaps.
For eg: assume the spec contains 
```yaml
      ports:
       -  name: bond0
          layer2: false
          bonded: true
          ip_addresses:
            - address: "192.168.2.0/24"
              vxlan_ids: [1000]
              gateway: "192.168.2.1"
       -  name: eth0
          layer2: false
          bonded: true
          ip_addresses:
            - address: "192.168.3.0/24"
              vxlan_ids: [1001]
              gateway: "192.168.3.1"
```

Assume there are 6 machines in a cluster, so in each of the machine we will have to manage the IP Address assigned to bond0 and eth0, so what I was thinking is CAPP will create 2 configmaps, `<cluster-name>-bond0` and `<cluster-name>-eht0` which will manage allcation and de-allocation to IP addressed for each machine in a cluster.
Let me know if you can think of any better way ?

Reply:
> 2 configmaps, <cluster-name>-bond0 and <cluster-name>-eth0 which will manage allcation

A user should be able to inspect or change the bond0 and eth0 configuration in one pass, especially to avoid racing the controller as it picks up and applies changes.

> address: "192.168.3.0/24"

For illustrative purposes, as in the snippet above, let's assume the node address is not `.0` (the network address) or `.1` (the gateway address)

> manage allcation and de-allocation to IP addressed for each machine in a cluster.

A per-cluster config map seems reasonable. What should it contain? 

Which values should are stored in PacketCluster CRDs instead? https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api-provider-packet/infrastructure.cluster.x-k8s.io/PacketCluster/v1beta1@v0.7.2

Perhaps something like this?
```yaml
kind: PacketCluster
spec:
  ...
  networks:
  - vlan: uuid
     ip_addresses: ["192.168.3.0/24"]
     assignment: cluster-api
     gateway: "192.168.3.1" # a default gateway
     routes:
     - destination: "192.168.4.0/24"
       gateway:  "192.168.3.2"       
  - vxlan: 1234
     assignment: dhcp
```


I added `dhcp` to illustrate that the PacketCluster may not need to track every assignment. This is a reasonable alternative in the BYO-Infrastructure approach, we assume the networks will already exist. We assume if DHCP is going to manage address assignments then it will already be present. 

Perhaps in a future iteration, we would want to provide details on how the DHCP services should be configured, what extra flags the service should pass. These considerations would take us down the path of PacketCluster managing network infrastructure (potentially bastion nodes). Let's ignore that for now. 

We may also want to allow the controller to assign multiple IP addresses to each node from the given range.  "32" vs "29" could be another parameter in the network definition to identify the per-node network assignment size. ("32": each PacketMachine would get one address in the range, "29": each PacketMachine would get eight addresses in the range)

```yaml
kind: PacketCluster
spec:
  networks:
  - vlan: uuid
    ip_addresses: ["192.168.3.0/24"]
    assignment: cluster-api
    gateway: "192.168.3.1"
    subnetSize: /29  # Specifies the subnet size for per-machine IP assignments
    routes:
    - destination: "192.168.4.0/24"
      gateway: "192.168.3.2"
```


And then perhaps the configmap looks something like this?

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-network-config   # Cluster1234
  namespace: default
data:
  networks:
  - vlan: {uuid} # this is our lookup against the PacketCluster object.. 
     # what networks defined with vxlan instead of vlan? vxlans (1234) within a project are not unique by themselves, they are project + metro scoped.
     ip_addresses: 
     - network: 192.168.3.0/24
        assignments:
        - address: 192.168.3.1
           machine: {packetmachine uuid ?}
           # or maybe device: {equinix metal API device uuid ?}
        - 
```

What other cluster scoped information is needed? 
Do we need to track a cluster-wide machine port configuration? (all machines should be in hybrid-bonded? hybrid-unbonded?)

Reply:
I don't think the ConfigMap would need to track details at the port level per machine; those values would be in the PacketMachine definition.

Reply:
I think I'm coming around to where you started. The PacketMachineTemplate, and its controller, may be where the ConfigMap needs to be managed. At this level, ports are relevant. In a hybrid-unbonded configuration, bond1 or eth1 may have VLAN assignments that are different than VLANs assigned to bond0.

Reply:
@displague does the subnetSize: /29  # Specifies the subnet size for per-machine IP assignments 
here specifies the max. amount of IPs a machine can get out of total range ? for eg: here /29 would mean 8 IPs out of total range ?